src/main/java/servlets/UploadReceiptServlet.java

```diff
@@ -0,0 +1,176 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet with a GET handler that creates a URL that uploads a receipt image to Blobstore and
+ * a POST handler that extracts data from the image and inserts it into Datastore.
+ */
+@WebServlet("/upload-receipt")
+public class UploadReceiptServlet extends HttpServlet {
+  // Matches JPEG image filenames.
+  private static final Pattern validFilename = Pattern.compile("([^\\s]+(\\.(?i)(jpe?g))$)");
+  // Logs to System.err by default.
+  private static final Logger logger = Logger.getLogger(UploadReceiptServlet.class.getName());
+
+  /**
+   * Creates a URL that uploads the receipt image to Blobstore when the user submits the upload
+   * form. After Blobstore handles the parsing, storing, and hosting of the image, the form
+   * data and a URL where the image can be accessed is forwarded to this servlet in a POST
+   * request.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstoreService.createUploadUrl("/upload-receipt");
```

Comment:
Do we want to set UploadOptions (e.g., https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/blobstore/UploadOptions.html#maxUploadSizeBytesPerBlob-long-)

Reply:
Added UploadOptions for a 5 MB maximum